pkg/util/workqueue/delaying_queue.go

```diff
@@ -0,0 +1,226 @@
+
+	insertionIndex := sort.Search(len(q.waitingForAdd), func(i int) bool {
+		return waitEntry.readyAt.Before(q.waitingForAdd[i].readyAt)
+	})
+
+	tail := q.waitingForAdd[insertionIndex:]
+	q.waitingForAdd = append(make([]waitFor, 0, len(q.waitingForAdd)+1), q.waitingForAdd[:insertionIndex]...)
+	q.waitingForAdd = append(q.waitingForAdd, waitEntry)
+	q.waitingForAdd = append(q.waitingForAdd, tail...)
+
+	q.notifyAt(waitEntry.readyAt)
+}
+
+// maxWait keeps a max bound on the wait time.  It's just insurance against weird things happening.
+// Checking the queue every 10 seconds isn't expensive and we know that we'll never end up with an
+// expired item sitting for more than 10 seconds.
+const maxWait = 10 * time.Second
+
+// waitingLoop runs until the workqueue is shutdown and keeps a check on the list of items to be added.
+func (q *delayingType) waitingLoop() {
+	defer utilruntime.HandleCrash()
+
+	for {
+		if q.shuttingDown {
+			return
+		}
+
+		func() {
+			q.waitingCond.L.Lock()
+			defer q.waitingCond.L.Unlock()
+			q.waitingCond.Wait()
+
+			if q.shuttingDown {
+				return
+			}
+
+			q.waitingLock.Lock()
+			defer q.waitingLock.Unlock()
+
+			nextReadyCheck := time.Time{}
+			itemsAdded := 0
+
+			for _, queuedItem := range q.waitingForAdd {
+				nextReadyCheck = queuedItem.readyAt
+				if queuedItem.readyAt.After(q.clock.Now()) {
+					break
+				}
+				q.Type.Add(queuedItem.data)
+				itemsAdded++
+			}
+
+			switch itemsAdded {
+			case 0:
+				// no change
+			case len(q.waitingForAdd):
+				// consumed everything
+				q.waitingForAdd = make([]waitFor, 0, len(q.waitingForAdd))
+
+			default:
+				// consumed some
+				q.waitingForAdd = q.waitingForAdd[itemsAdded:]
+
+				if len(q.waitingForAdd) > 0 {
+					q.notifyAt(nextReadyCheck)
+				}
+			}
+		}()
+	}
+}
+
+// heartbeat forces a check every maxWait seconds
+func (q *delayingType) heartbeat() {
+	defer utilruntime.HandleCrash()
+
+	for {
+		if q.shuttingDown {
+			return
+		}
+
+		ch := q.clock.After(maxWait)
+		<-ch
+		q.waitingCond.Broadcast()
+	}
+}
+
+// clearNextCheckTimeIf resets the nextCheckTime if it matches the expected value to ensure that the subsequent notification will take effect.
+func (q *delayingType) clearNextCheckTimeIf(nextReadyCheck time.Time) {
+	q.nextCheckLock.Lock()
+	defer q.nextCheckLock.Unlock()
+
+	if q.nextCheckTime != nil && *q.nextCheckTime == nextReadyCheck {
+		q.nextCheckTime = nil
+	}
+}
+
+// notifyAt: if the requested nextReadyCheck is sooner than the current check, then a new go func is
+// spawned to notify the condition that the waitingLoop is waiting for after the time is up.  The previous go func
+// is cancelled
+func (q *delayingType) notifyAt(nextReadyCheck time.Time) {
+	q.nextCheckLock.Lock()
+	defer q.nextCheckLock.Unlock()
+
+	now := q.clock.Now()
+	if (q.nextCheckTime != nil && (nextReadyCheck.After(*q.nextCheckTime) || nextReadyCheck == *q.nextCheckTime)) || nextReadyCheck.Before(now) {
```

Comment:
@liggitt this is intended to stop one-per `AddAfter`.  If you add later in the list before the previous go routine has fired, then you don't get a goroutine.  So you get elimination of unnecessaries and cancellation of expired ones.